content/2021-06-11-make-your-life-easier.md

```diff
@@ -0,0 +1,156 @@
+working on them. Here are some things to remember:
+
+- There's no such thing as a zero-downtime atomic deploy. You can choose
+  downtime if you want, but choosing a safe, non-atomic deployment pattern is
+  much easier.
+- Your system's topology is a graph. The guidance for graph data structures
+  above applies here too. If you have cycles in your graph of services you
+  will regret it.
+- There's no such thing as a reliable network. It will fail, so don't use it
+  unless you need it. If you make your services very "micro" then you are
+  probably hitting the network more than you want.
+- Eventual consistency is often good enough. Use it enthusiastically.
+- Exactly-once delivery of messages is hard. At-least-once delivery of
+  messages that are idempotent is much, much easier.
+- Don't share data stores between different services. Have one, canonical
+  owning service for every piece of data. Other services may cache the data,
+  but they need to defer to the owning service.
+- Use centralised logging and metrics systems from the beginning. Use a tool
+  like [Honeycomb](https://www.honeycomb.io/) if you can. You will learn
+  things about your system every day.
+
+# Everything should have limits
+
+<blockquote class="twitter-tweet"><p lang="en" dir="ltr">Every new feature should have a limit of 5</p>&mdash; Gordon Syme (@gordon_syme) <a href="https://twitter.com/gordon_syme/status/1400036732763049986?ref_src=twsrc%5Etfw">June 2, 2021</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
+
+This is somewhat tongue-in-cheek but the point still stands. It's relatively
+easy to relax a limitation, but if you don't have one to begin with you will
+have to negotiate to introduce it.
+
+Types of limits worth thinking about are:
+
+- Request rate limits. An accidental denial of service by an overzealous
+  customer is embarrassing for everyone.
+- Request size limits. These apply on both input and output, for example
+  limiting the size of a text field being inserted and the number of records
+  that can be fetched at once.
+- Pagination. If you return a sequence of anything it should be paginated and
+  there should be a limit on the maximum size of the page.
+- Application-specific limits. For example, if you're making an online store
+  limit the number of product categories and limit the number of products per
+  category. Everywhere you see a list, a one-to-many or a many-to-many
+  relationship you should attempt to limit it.
+- Time-based retention limits for data at rest. Don't promise to store things
+  forever. If you don't want to implement the deletion immediately, hide it so
+  that the customers are used to the idea that data goes away.
+- Security limits. For example browser session lengths, failed login attempts,
+  etc.
+
+# Trust your gut
+
+There are many other rules of thumb that I use in my day-to-day work, but it's
```

Comment:
I cannot remember who said it, but "Experience is advanced pattern recognition" is super true.

Reply:
:nods-intensifies-more: